README.md
# electrical-meter ![Docker Hub image](https://github.com/Froschie/electrical-meter/workflows/Docker%20Hub%20image/badge.svg)

Script to read SML messages from electrical meter and save it to InfluxDB.  

Electric meters offer various informations via SML (Smart Message Language) IR interface.  
The [SML](https://de.wikipedia.org/wiki/Smart_Message_Language) protocol is using specified [OBIS Numbers](https://de.wikipedia.org/wiki/OBIS-Kennzahlen) which contain informations like actual power consumption or total energy usage.  
The messages can be captured by a simply IR reader with USB connection (search for "smartmeter Optokopf").  

My script is capturing all the SML messages and by using Python [smllib](https://github.com/spacemanspiff2007/SmlLib) Library decoding it.

The script was tested on a ISKRA MT631 electric meter with extended informations enabled (PIN from power operator necessary).

*Note: the smart meters (at least in Germany) are by default not providing detailed informations! This needs to be manually enabled. To do this please request the necessary PIN from your Energy Network Provider and then set the Energy Meter accordingly. Without extended informations enabled current power consumption is not transmitted and counters only provide Wh values with last 3x digits set to 0 (virtually reporting kWh only).*


## Part List
USB IR Reader <50€


## Start Docker Container  

Pull latest Image:  
`docker pull froschie/sml-electrical-meter:latest`  

Start Container to only query values:  
```
docker run -it --name temp --rm \
 --device /dev/ttyUSB0 \
 -e device="/dev/ttyUSB0" \
 -e TZ=Europe/Berlin \
 -e write="0" \
 -e interval=1 \
 -e influx_ip="" \
 froschie/sml-electrical-meter
```
*Note: please adapt the parameters as needed especially the device parameters!*  

Example Output:
```
2021-12-30 13:19:05.00203    INFO:      Actual Time: 2021-12-30 13:19:05.203273 waiting for: 2021-12-30 13:19:06
2021-12-30 13:19:07.00144    INFO:      96.50.1: ISK 96.1.0: 1234567890 1.8.0: 1234567.1Wh 2.8.0: 123456.1Wh 16.7.0: -50W
2021-12-30 13:19:08.00150    INFO:      96.50.1: ISK 96.1.0: 1234567890 1.8.0: 1234567.2Wh 2.8.0: 123456.2Wh 16.7.0: -25W
2021-12-30 13:19:09.00141    INFO:      96.50.1: ISK 96.1.0: 1234567890 1.8.0: 1234567.3Wh 2.8.0: 123456.3Wh 16.7.0: 0W
2021-12-30 13:19:10.00149    INFO:      96.50.1: ISK 96.1.0: 1234567890 1.8.0: 1234567.4Wh 2.8.0: 123456.4Wh 16.7.0: 25W
2021-12-30 13:19:11.00141    INFO:      96.50.1: ISK 96.1.0: 1234567890 1.8.0: 1234567.5Wh 2.8.0: 123456.5Wh 16.7.0: 50W
```

Negative Values mean generated power from photovoltaik is suplied to the grid. Positive values mean power is taken from grid.


## Start Docker Container via Docker-Compose  
```bash
curl -O https://raw.githubusercontent.com/Froschie/electrical-meter/master/docker-compose.yaml
vi docker-compose.yaml
docker-compose up -d
```

Example Docker Compose file:
```yaml
  electric-meter-query:
    image: electric-meter
    container_name: electric-meter-query
    environment:
      - influx_ip=192.168.1.3
      - influx_port=8086
      - influx_user=<user>
      - influx_pw=<pw2>
      - influx_db=<db>
      - interval=300
      - write=0
      - device=/dev/ttyUSB0
      - TZ=Europe/Berlin
      - log=INFO
      - obis=SHORT
      - unit=INT
    devices:
      - /dev/ttyUSB0
    logging: 
      options:
        max-size: "5M"
        max-file: "10"
    restart: unless-stopped
```

*Note: please adapt the parameters as needed!*


## Create a Docker Container

```bash
mkdir electric-meter
cd electric-meter/
curl -O https://raw.githubusercontent.com/Froschie/electrical-meter/master/Dockerfile
curl -O https://raw.githubusercontent.com/Froschie/electrical-meter/master/query_electric_meter.py
curl -O https://raw.githubusercontent.com/Froschie/electrical-meter/master/influx_fields.json
curl -O https://raw.githubusercontent.com/Froschie/electrical-meter/master/obis_code_list.json
curl -O https://raw.githubusercontent.com/Froschie/electrical-meter/master/obis_unit_list.json
docker build --tag electrical-meter .
```

## Run Script without Docker (query only)

With latest Python Version execute:

```bash
pip install influxdb pyserial smllib
python query_electric_meter.py --device=/dev/ttyUSB0 --interval=1
```

To execute Test it in a Python Docker container: 

```bash
git clone https://github.com/Froschie/electrical-meter.git
cd electrical-meter
docker run -it --name temp --rm --device /dev/ttyUSB0 -v $(pwd):/mnt/em python bash
pip install influxdb pyserial smllib
cd /mnt/em
python query_electric_meter.py --device=/dev/ttyUSB0 --interval=1
```

## Script Options

| Option | Description |
| --- | --- |
| device | USB Device path, default: `/dev/ttyUSB0` |
| interval | Interval displaying the SML message, default `60`, any integer from "1" is allowed |
| obis | OBIS Code Format, default `SHORT` (e.g. 1.8.0), possible values: "SHORT", "LONG"(1-0.1.8.0), "NAME" (Positive active energy) |
| write | Enable writing to InluxDB usefull for debugging or initial testing, default `0` (Disabled), possible values: "0", "1" (enabled) |
| influx_ip | InfluxDB IP |
| influx_port | InfluxDB Port |
| influx_user | InfluxDB Username |
| influx_pw | InfluxDB Password |
| influx_db | Influx DB Name |
| unit | Unit Value Format when writing to InfluxDB, default: `INT` (Integer), possible values: "INT", "FLOAT" |
| log | Output Level, default: `INFO`, possible values: "DEBUG", "INFO", "WARNING", "ERROR" |

## Configuration files

The script uses 3x config files which can be modified if needed. The current files are based on the fields my Energy Meter is offering.

This file defines which fields are written to InfluxDB, `influx_fields.json`:
```json
{
    "tags": {
        "96.50.1": "device",
        "96.1.0": "serial"
    },
    "fields": {
        "1.8.0": "consumption",
        "2.8.0": "supply"
    }
}
```

This file defines the names for the OBIS codes, `obis_code_list.json`:
```json
{
    "96.50.1": "Manufacturer",
    "1.8.0": "Positive active energy",
    "2.8.0": "Negative active energy",
    "96.1.0": "Serial Number",
    "16.7.0": "Sum active instantaneous power"
}
```

This file defines the units of the OBIS values, `obis_unit_list.json`:
```json
{
    "30": "Wh",
    "27": "W"
}
```

To use a customized configuration file,  start Container with mapping the file to root folder:
```bash
docker run -it --name temp --rm \
 --device /dev/ttyUSB0 \
 -e device="/dev/ttyUSB0" \
 -e TZ=Europe/Berlin \
 -e write="0" \
 -e interval=1 \
 -e influx_ip="" \
 -v $(pwd)/obis_code_list.json:/obis_code_list.json \
 froschie/sml-electrical-meter
```

Or via Docker Compose file:
```yaml
  electric-meter-query:
  ...
    volumes:
      - /directory/obis_code_list.json:/obis_code_list.json
  ...
```


CHANGELOG
v4.1 - 26.01.2021
- open USB port only when reading data

v4.01 - 06.01.2021
- change Exception logging to debug and add seperate InfluxDB exception handling

v4.0 - 30.12.2021
- complete rewrite using smllib python library
- added configuration files for flexible OBIS code usage


get_electric_meter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
 
import time
import serial
from threading import Timer
import sys
import os
from datetime import datetime, timedelta
from influxdb import InfluxDBClient
import argparse

parser=argparse.ArgumentParser(
    description='''Script for Query SML Values from Electric Meter.''')
parser.add_argument('--device', type=str, required=False, default="/dev/ttyUSB0", help='Path to SML Reader Device.')
parser.add_argument('--influx_ip', type=str, required=True, default="", help='IP of the Influx DB Server.')
parser.add_argument('--influx_port', type=str, required=True, default="", help='Port of the Influx DB Server.')
parser.add_argument('--influx_user', type=str, required=True, default="", help='User of the Influx DB Server.')
parser.add_argument('--influx_pw', type=str, required=True, default="", help='Password of the Influx DB Server.')
parser.add_argument('--influx_db', type=str, required=True, default="", help='DB name of the Influx DB Server.')
parser.add_argument('--write', type=int, required=False, default=1, choices=[0, 1], help='Specify if Data should be written to InfluxDB or not.')
parser.add_argument('--interval', type=int, required=False, default=60, help='Query interval in "s".')
parser.add_argument('--maxruntime', type=int, required=False, default=86400, help='Max script runtime in "s".')
args=parser.parse_args()

# Time Rounding Function
def ceil_time(ct, delta):
    return ct + (datetime.min - ct) % delta

starttime = datetime.now()
now = datetime.now()
new_time = ceil_time(now, timedelta(seconds=int(args.interval)))
print(now, "Actual Time:", now, "waiting for:", new_time)    

# Wait for Full Minute / Half Minute
while now < new_time:
    time.sleep(0.5)
    now = datetime.now()

mystring = b""
crc16_x25_table = [
    0x0000, 0x1189, 0x2312, 0x329B, 0x4624, 0x57AD,    0x6536, 0x74BF,
    0x8C48, 0x9DC1, 0xAF5A, 0xBED3, 0xCA6C, 0xDBE5, 0xE97E, 0xF8F7,
    0x1081, 0x0108,    0x3393, 0x221A, 0x56A5, 0x472C, 0x75B7, 0x643E,
    0x9CC9, 0x8D40, 0xBFDB, 0xAE52, 0xDAED, 0xCB64,    0xF9FF, 0xE876,
    0x2102, 0x308B, 0x0210, 0x1399,    0x6726, 0x76AF, 0x4434, 0x55BD,
    0xAD4A, 0xBCC3,    0x8E58, 0x9FD1, 0xEB6E, 0xFAE7, 0xC87C, 0xD9F5,
    0x3183, 0x200A, 0x1291, 0x0318, 0x77A7, 0x662E,    0x54B5, 0x453C,
    0xBDCB, 0xAC42, 0x9ED9, 0x8F50,    0xFBEF, 0xEA66, 0xD8FD, 0xC974,
    0x4204, 0x538D,    0x6116, 0x709F, 0x0420, 0x15A9, 0x2732, 0x36BB,
    0xCE4C, 0xDFC5, 0xED5E, 0xFCD7, 0x8868, 0x99E1,    0xAB7A, 0xBAF3,
    0x5285, 0x430C, 0x7197, 0x601E,    0x14A1, 0x0528, 0x37B3, 0x263A,
    0xDECD, 0xCF44,    0xFDDF, 0xEC56, 0x98E9, 0x8960, 0xBBFB, 0xAA72,
    0x6306, 0x728F, 0x4014, 0x519D, 0x2522, 0x34AB,    0x0630, 0x17B9,
    0xEF4E, 0xFEC7, 0xCC5C, 0xDDD5,    0xA96A, 0xB8E3, 0x8A78, 0x9BF1,
    0x7387, 0x620E,    0x5095, 0x411C, 0x35A3, 0x242A, 0x16B1, 0x0738,
    0xFFCF, 0xEE46, 0xDCDD, 0xCD54, 0xB9EB, 0xA862,    0x9AF9, 0x8B70,
    0x8408, 0x9581, 0xA71A, 0xB693,    0xC22C, 0xD3A5, 0xE13E, 0xF0B7,
    0x0840, 0x19C9,    0x2B52, 0x3ADB, 0x4E64, 0x5FED, 0x6D76, 0x7CFF,
    0x9489, 0x8500, 0xB79B, 0xA612, 0xD2AD, 0xC324,    0xF1BF, 0xE036,
    0x18C1, 0x0948, 0x3BD3, 0x2A5A,    0x5EE5, 0x4F6C, 0x7DF7, 0x6C7E,
    0xA50A, 0xB483,    0x8618, 0x9791, 0xE32E, 0xF2A7, 0xC03C, 0xD1B5,
    0x2942, 0x38CB, 0x0A50, 0x1BD9, 0x6F66, 0x7EEF,    0x4C74, 0x5DFD,
    0xB58B, 0xA402, 0x9699, 0x8710,    0xF3AF, 0xE226, 0xD0BD, 0xC134,
    0x39C3, 0x284A,    0x1AD1, 0x0B58, 0x7FE7, 0x6E6E, 0x5CF5, 0x4D7C,
    0xC60C, 0xD785, 0xE51E, 0xF497, 0x8028, 0x91A1,    0xA33A, 0xB2B3,
    0x4A44, 0x5BCD, 0x6956, 0x78DF,    0x0C60, 0x1DE9, 0x2F72, 0x3EFB,
    0xD68D, 0xC704,    0xF59F, 0xE416, 0x90A9, 0x8120, 0xB3BB, 0xA232,
    0x5AC5, 0x4B4C, 0x79D7, 0x685E, 0x1CE1, 0x0D68,    0x3FF3, 0x2E7A,
    0xE70E, 0xF687, 0xC41C, 0xD595,    0xA12A, 0xB0A3, 0x8238, 0x93B1,
    0x6B46, 0x7ACF,    0x4854, 0x59DD, 0x2D62, 0x3CEB, 0x0E70, 0x1FF9,
    0xF78F, 0xE606, 0xD49D, 0xC514, 0xB1AB, 0xA022,    0x92B9, 0x8330,
    0x7BC7, 0x6A4E, 0x58D5, 0x495C,    0x3DE3, 0x2C6A, 0x1EF1, 0x0F78]

class Watchdog_timer:
    def __init__(self, timeout, userHandler=None):
        self.timeout = timeout
        self.handler = userHandler if userHandler is not None else self.defaultHandler
        self.timer = Timer(self.timeout, self.handler)
        self.timer.start()
 
    def reset(self):
        self.timer.cancel()
        self.timer = Timer(self.timeout, self.handler)
        self.timer.start()
 
    def stop(self):
        self.timer.cancel()
 
    def defaultHandler(self):
        raise self

def crc16_x25(Buffer):
    crcsum = 0xffff
    global crc16_x25_table
    Buffer = [b'%c' % i for i in Buffer]
    for byte in Buffer:
        crcsum = crc16_x25_table[(ord(byte) ^ crcsum) & 0xff] ^ (crcsum >> 8 & 0xff)
    crcsum ^= 0xffff
    return crcsum

str_consumption = "77070100010800ff"
str_supply = "77070100020800ff"
str_actpower = "77070100100700ff"

def twos_comp(val, bits):
    if (val & (1 << (bits - 1))) != 0:
        val = val - (1 << bits)
    return val

def watchdogtimer_ovf():
    global mystring
    global readLoop
    message = mystring[0:-2]
    crc_calc = crc16_x25(message)
    crc_rx = int(str(hex(mystring[-1]))[2:]+str(hex(mystring[-2]))[2:] ,16)
    if crc_rx == crc_calc:
        if message[0:8] == b'\x1b\x1b\x1b\x1b\x01\x01\x01\x01':
            message = mystring.hex()
            counter_consumption = 0
            counter_supply = 0
            counter_actpower = 0
            if message.find(str_consumption):
                start_consumption = message.find(str_consumption)+(len(str_consumption))
                end_consumption = message[start_consumption:].find("0177")
                value_consumption = message[start_consumption:start_consumption+end_consumption+2]
                lengh_consumption_str = int(value_consumption[value_consumption.find("621e52ff")+(len("621e52ff")+1):value_consumption.find("621e52ff")+(len("621e52ff")+2)])-1
                counter_consumption = int(value_consumption[value_consumption.find("621e52ff")+(len("621e52ff")+2):value_consumption.find("621e52ff")+(len("621e52ff")+2)+lengh_consumption_str*2], 16)/10
            if message.find(str_supply):
                start_supply = message.find(str_supply)+(len(str_supply))
                end_supply = message[start_supply:].find("0177")
                value_supply = message[start_supply:start_supply+end_supply+2]
                lengh_supply_str = int(value_supply[value_supply.find("621e52ff")+(len("621e52ff")+1):value_supply.find("621e52ff")+(len("621e52ff")+2)])-1
                counter_supply = int(value_supply[value_supply.find("621e52ff")+(len("621e52ff")+2):value_supply.find("621e52ff")+(len("621e52ff")+2)+lengh_supply_str*2], 16)/10
            if message.find(str_actpower):
                start_actpower = message.find(str_actpower)+(len(str_actpower))
                end_actpower = message[start_actpower:].find("010101")
                value_actpower = message[start_actpower:start_actpower+end_actpower+2]
                lengh_actpower_str = int(value_actpower[value_actpower.find("621b5200")+(len("621b5200")+1):value_actpower.find("621b5200")+(len("621b5200")+2)])-1
                counter_actpower = twos_comp(int(value_actpower[value_actpower.find("621b5200")+(len("621b5200")+2):value_actpower.find("621b5200")+(len("621b5200")+2)+lengh_actpower_str*2], 16), lengh_actpower_str*8)
            print(datetime.now(), "Consumption: ", counter_consumption, "Wh, ", "Supply: ", counter_supply, "Wh, ", "Actual Power: ", counter_actpower, "W")
            if counter_consumption != "---" and counter_supply != "---" and args.write == 1:
                measurement_list = {'consumption': counter_consumption, 'supply': counter_supply}
                temp_body = {
                    "measurement": "electric_meter",
                    "time": datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ'),
                    "fields": {}
                }
                for measurement in measurement_list:
                    if measurement_list[measurement] > 0:
                        temp_body['fields'][measurement] = int(measurement_list[measurement])
                if len(temp_body['fields']) > 0:
                    json_body = [temp_body]
                    #print(datetime.now(), "InfluxDB write data DEBUG:" + str(json_body))
                    client = InfluxDBClient(host=args.influx_ip, port=args.influx_port, username=args.influx_user, password=args.influx_pw)
                    client.switch_database(args.influx_db)
                    influx_result = client.write_points(json_body)
                    if influx_result:
                        print(datetime.now(), "InfluxDB write data successfull:" + str(json_body))
                    else:
                        print(datetime.now(), "InfluxDB write data FAILED:" + str(json_body))
                        print(influx_result)
                    client.close()
                else:
                    print(datetime.now(), "InfluxDB write skipped, no valid data found:", measurement_list)
            readLoop = False
            watchdog.stop()
            mystring=b""
        else:
            mystring=b""
            watchdog.stop()
    else:
        mystring=b""
        watchdog.stop()

try:
    print(datetime.now(), "Query Electric Meter Values started with interval {}...".format(args.interval))
    while True:
        try:
            my_tty = serial.Serial(port=args.device, baudrate = 9600, parity =serial.PARITY_NONE, stopbits=serial.STOPBITS_ONE, bytesize=serial.EIGHTBITS, timeout=0)
            my_tty.close()
            my_tty.open()
        except Exception as e:
            sys.stdout.write("seriel port could not be opened:\n" + str(e) + "\n\n")
            time.sleep(1)
            exit()
        try:
            my_tty.reset_input_buffer()
            my_tty.reset_output_buffer()
            watchdog = Watchdog_timer(0.1, watchdogtimer_ovf)
            watchdog.stop()
            readLoop = True
            while readLoop:
                while my_tty.in_waiting > 0:
                    mystring += my_tty.read()
                    watchdog.reset()
        except KeyboardInterrupt:
            my_tty.close()
            sys.stdout.write("\nscript manually exited!\n")

        runtime = datetime.now() - starttime
        if runtime.total_seconds() > args.maxruntime:
            sys.exit(0)
        time.sleep(args.interval - ((time.time() - new_time.timestamp()) % args.interval))
except KeyboardInterrupt:
    print("Script aborted...")
finally:
    print("Script exited...")


get_electric_meter_start.py
#!/usr/bin/env python3
import os
import subprocess
from datetime import datetime, timedelta
import time
influx_ip = os.environ['influx_ip']
influx_port = os.environ['influx_port']
influx_user = os.environ['influx_user']
influx_pw = os.environ['influx_pw']
influx_db = os.environ['influx_db']
device = os.environ['device']
write = os.environ['write']
interval = float(os.environ['interval'])

cmd = '/get_electric_meter.py --influx_ip="%s" --influx_port="%s" --influx_user="%s" --influx_pw="%s" --influx_db="%s" --device="%s" --write="%s"' % (influx_ip, influx_port, influx_user, influx_pw, influx_db, device, write)

# Time Rounding Function
def ceil_time(ct, delta):
    return ct + (datetime.min - ct) % delta

now = datetime.now()
new_time = ceil_time(now, timedelta(seconds=int(interval)))
print(now, "Actual Time:", now, "waiting for:", new_time)    

# Wait for Full Minute / Half Minute
while now < new_time:
    time.sleep(0.5)
    now = datetime.now()

try:
    while True:
        print(datetime.now(), "Query Electric Meter Values...")
        p = subprocess.Popen(cmd, shell=True)
        p.wait()
        p.terminate()
        time.sleep(interval - ((time.time() - new_time.timestamp()) % interval))
except KeyboardInterrupt:
    print("Script aborted...")
finally:
    print("Script exited...")


query_electric_meter.py
# pip install influxdb pyserial smllib
import serial
from smllib import SmlStreamReader
import logging
import sys
import os
import json
import time
from datetime import datetime, timedelta
from influxdb import InfluxDBClient
import argparse

parser=argparse.ArgumentParser(
    description='''Script for Query SML Values from Electric Meter.''')
parser.add_argument('--device', type=str, required=False, default="/dev/ttyUSB0", help='Path to SML Reader Device.')
parser.add_argument('--influx_ip', type=str, required=False, default="", help='IP of the Influx DB Server.')
parser.add_argument('--influx_port', type=str, required=False, default="", help='Port of the Influx DB Server.')
parser.add_argument('--influx_user', type=str, required=False, default="", help='User of the Influx DB Server.')
parser.add_argument('--influx_pw', type=str, required=False, default="", help='Password of the Influx DB Server.')
parser.add_argument('--influx_db', type=str, required=False, default="", help='DB name of the Influx DB Server.')
parser.add_argument('--write', type=int, required=False, default=0, choices=[0, 1], help='Specify if Data should be written to InfluxDB or not.')
parser.add_argument('--interval', type=int, required=False, default=60, help='Query interval in "s".')
parser.add_argument('--unit', type=str, required=False, default="INT", choices=["INT", "FLOAT"], help='Specify the used value format for InfluxDB.')
parser.add_argument('--obis', type=str, required=False, default="SHORT", choices=["SHORT", "LONG", "NAME"], help='Specify in which format the OBIS Codes should be displayed.')
parser.add_argument('--log', type=str, required=False, default="INFO", choices=["DEBUG", "INFO", "WARNING", "ERROR"], help='Specify Log output.')
args=parser.parse_args()

# Log Output configuration
logging.basicConfig(level=getattr(logging, args.log), format='%(asctime)s.%(msecs)05d %(levelname)07s:\t%(message)s', datefmt='%Y-%m-%d %H:%M:%S')
log = logging.getLogger(__name__)

# Time Rounding Function
def ceil_time(ct, delta):
    return ct + (datetime.min - ct) % delta

starttime = datetime.now()
now = datetime.now()
new_time = ceil_time(now, timedelta(seconds=int(args.interval)))
log.info("Actual Time: " + str(now) + " waiting for: " + str(new_time))

# Wait for Full Minute / Half Minute
while now < new_time:
    time.sleep(0.5)
    now = datetime.now()

my_tty = serial.Serial(port=args.device, baudrate=9600, parity=serial.PARITY_NONE, stopbits=serial.STOPBITS_ONE, bytesize=serial.EIGHTBITS, timeout=0)
my_tty.close()
my_tty.open()
start = '1b1b1b1b01010101'
end = '1b1b1b1b1a'
data = ''

# Load OBIS Code List
try:
    for filename in ["obis_code_list.json", "/obis_code_list.json"]:
        if os.path.isfile(filename):
            with open(filename) as json_data_file:
                obis_code_list = json.load(json_data_file)
                break
except Exception as e:
    log.error("obis_code_list.json file could not be opened or is not valid JSON file!")
    log.error(e)
    time.sleep(60)
    sys.exit(1)

# Load OBIS Unit List
try:
    for filename in ["obis_unit_list.json", "/obis_unit_list.json"]:
        if os.path.isfile(filename):
            with open(filename) as json_data_file:
                obis_unit_list = json.load(json_data_file)
                break
except Exception as e:
    log.error("obis_unit_list.json file could not be opened or is not valid JSON file!")
    log.error(e)
    time.sleep(60)
    sys.exit(1)

# Load Influx Definiton for Tags and Fields
try:
    for filename in ["influx_fields.json", "/influx_fields.json"]:
        if os.path.isfile(filename):
            with open(filename) as json_data_file:
                influx_fields = json.load(json_data_file)
                break
except Exception as e:
    log.error("influx_fields.json file could not be opened or is not valid JSON file!")
    log.error(e)
    time.sleep(60)
    sys.exit(1)

def obis_code(value, format):
    obis_A = int(value[0:2], 16)
    obis_B = int(value[2:4], 16)
    obis_C = int(value[4:6], 16)
    obis_D = int(value[6:8], 16)
    obis_E = int(value[8:10], 16)
    if format == "SHORT" or format == "NAME":
        obis_string = "{}.{}.{}".format(obis_C, obis_D, obis_E)
    if format == "NAME" and obis_string in obis_code_list:
        obis_string = obis_code_list[obis_string]
    if format == "LONG":
        obis_string = "{}-{}:{}.{}.{}".format(obis_A, obis_B, obis_C, obis_D, obis_E)
    return obis_string

def obis_unit(unit):
    if str(unit) in obis_unit_list:
        return obis_unit_list[str(unit)]
    else:
        return ""

def obis_value(value, code):
    if code == "96.1.0":
        return str(int(value[-10:], 16)).rjust(10, '0')
    else:
        return str(value)

try:
    if args.influx_ip and args.influx_port and args.influx_user and args.influx_pw and args.influx_db:
        client = InfluxDBClient(host=args.influx_ip, port=args.influx_port, username=args.influx_user, password=args.influx_pw)
        client.switch_database(args.influx_db)
    while True:
        try:
            char = my_tty.read()
            data = data + char.hex()
            pos = data.find(end)
            if (pos != -1):
                for x in range(0, 3):
                    char = my_tty.read()
                    data = data + char.hex() 
                log.debug(data)
                stream = SmlStreamReader()
                stream.add(bytes.fromhex(data))
                sml_frame = stream.get_frame()
                if sml_frame is None:
                    continue
                obis_values = sml_frame.get_obis()
                text = ""
                if args.influx_ip and args.influx_port and args.influx_user and args.influx_pw and args.influx_db:
                    temp_body = {
                        "measurement": "electric_meter",
                        "tags": {},
                        "time": datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ'),
                        "fields": {}
                    }
                for item in obis_values:
                    code_short = obis_code(item.obis, "SHORT")
                    value = obis_value(item.get_value(), code_short)
                    if args.influx_ip and args.influx_port and args.influx_user and args.influx_pw and args.influx_db:
                        if code_short in influx_fields['tags']:
                            temp_body['tags'][influx_fields['tags'][code_short]] = value
                        if code_short in influx_fields['fields']:
                            points = client.query("SELECT last({}) FROM electric_meter WHERE serial='{}'".format(influx_fields['fields'][code_short], temp_body['tags']['serial'])).get_points()
                            last_value = 0
                            for point in points:
                                last_value = point['last']
                            if float(value) >= float(last_value) or points is None:
                                if args.unit == "INT":
                                    temp_body['fields'][influx_fields['fields'][code_short]] = int(float(value))
                                else:
                                    temp_body['fields'][influx_fields['fields'][code_short]] = float(value)
                            else:
                                log.warning("Point value below last recoded one! ({}: {})".format(influx_fields['fields'][code_short], value))
                    text += obis_code(item.obis, args.obis) + ": " + value + obis_unit(item.unit) + " "
                log.info(text)
                if args.influx_ip and args.influx_port and args.influx_user and args.influx_pw and args.influx_db:
                    try:
                        log.debug(temp_body)
                        json_body = [temp_body]
                        if args.write == 1:
                            influx_result = client.write_points(json_body)
                            if influx_result:
                                log.debug("InfluxDB write data successfull:" + str(json_body))
                            else:
                                log.error("InfluxDB write data FAILED:" + str(json_body))
                                log.error(influx_result)
                        #client.close()
                    except Exception as e:
                        log.error("InfluxDB writing failed: {}".format(e))
                my_tty.close()
                time.sleep(args.interval - ((time.time() - new_time.timestamp()) % args.interval))
                my_tty.open()
                data = ''
        except KeyboardInterrupt:
            log.warning("Script aborted...")
            break
        except Exception as e:
            log.debug("Something went wrong: {}".format(e))
            continue
except KeyboardInterrupt:
    log.warning("Script aborted...")
finally:
    log.info("Script ending...")
